### PR TITLE
feat: add restapi block

### PR DESCRIPTION
**Pull request by Aayush-Abhyarthi**
### Description

https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/145

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -215,3 +215,23 @@ data "ibm_database_connection" "database_connection" {
...
+  path          = "//${data.ibm_database_connection.database_connection.https[0].hosts[0].hostname}:${data.ibm_database_connection.database_connection.https[0].hosts[0].port}/_ml/trained_models/.elser_model_1?pretty"
+  create_method = "PUT"
+  data = jsonencode({
+    input = {
+      field_names = ["text_field"]
+    }
+  })
+}
+
+resource "restapi_object" "start_trained_model_deployment" {
```
Comment (daniel-butler-irl) [inline main.tf:230]: It was my understanding that to interact with the database you need to use the database internal credentials not the IBM Cloud IAM credentials that seem to be used here. Not sure if this is related to the pipeline failure.

**Comment by rajatagarwal-ibm** [comment]
@Aayush-Abhyarthi did you try to connect to the DB instance? You probably need DB credentials to connect to the DB instead of the IBM Cloud credentials, can you confirm, please?

**Comment by Aayush-Abhyarthi** [comment]
I added the admin user name and password in the ES connection url for authentication. But we are not passing any password for the admin user through our module while testing and this is the reason the pipeline is failing.
Could we add a random password in the pr_test.go file for testing the DA..? similar to what postgresql does? [here](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/blob/eecc97a300b09cee1e04ad9c67d3898dc726da93/tests/pr_test.go#L102C6-L102C28)
@daniel-butler-irl @rajatagarwal-ibm

**Comment by daniel-butler-irl** [comment]
The terraform is executed from the pipeline, I expect there are security rules set up blocking the traffic from the public internet. This is blocking your REST calls causing a timeout. @ocofaigh do we solve this anywhere else yet?

File: variables.tf [hunk 2]
```diff
@@ -242,3 +242,9 @@ variable "backup_crn" {
     error_message = "backup_crn must be null OR starts with 'crn:' and contains ':backup:'"
   }
 }
+
+variable "enable_elser_model" {
+  type        = bool
+  description = "Set it to true to install and start the elser model"
```
Comment (ocofaigh) [inline variables.tf:248]: elser model -> Elastic's Natural Language Processing model

Also Add this `[Learn more](https://cloud.ibm.com/docs/databases-for-elasticsearch?topic=databases-for-elasticsearch-elser-embeddings-elasticsearch)`

File: solutions/secure/version.tf [hunk 3]
```diff
@@ -7,5 +7,9 @@ terraform {
       source  = "IBM-Cloud/ibm"
       version = "1.66.0"
     }
+    restapi = {
+      source  = "Mastercard/restapi"
+      version = ">= 1.19.1"
```
Comment (ocofaigh) [inline solutions/secure/version.tf:12]: DAs should lock into exact provider version

File: solutions/secure/variables.tf [hunk 4]
```diff
@@ -171,3 +171,9 @@ variable "auto_scaling" {
   description = "The rules to allow the database to increase resources in response to usage. Only a single autoscaling block is allowed. Make sure you understand the effects of autoscaling, especially for production environments. See https://cloud.ibm.com/docs/databases-for-elasticsearch?topic=databases-for-elasticsearch-autoscaling&interface=cli#autoscaling-considerations in the IBM Cloud Docs."
   default     = null
 }
+
+variable "enable_elser_model" {
+  type        = bool
+  description = "Set it to true to install and start the elser model"
```
Comment (ocofaigh) [inline solutions/secure/variables.tf:177]: same as previous comment on variable description update

Not sure this should default to `true` ?

File: modules/fscloud/variables.tf [hunk 5]
```diff
@@ -177,3 +177,9 @@ variable "cbr_rules" {
   default     = []
   # Validation happens in the rule module
 }
+
+variable "enable_elser_model" {
+  type        = bool
+  description = "Set it to true to install and start the elser model"
```
Comment (ocofaigh) [inline modules/fscloud/variables.tf:183]: same as previous comment on variable description update

**Review by ocofaigh — CHANGES_REQUESTED** [review]
The test is failing because its trying to hit a private endpoint. You will need to update the test to run in schematics so you have access to the private endpoint

File: version.tf [hunk 6]
```diff
@@ -10,5 +10,9 @@ terraform {
       source  = "hashicorp/time"
       version = ">= 0.9.1"
     }
+    restapi = {
+      source  = "Mastercard/restapi"
+      version = ">= 1.19.1"
```
[the remark this answers is not in the document]
Comment (ocofaigh) [reply version.tf:15]: please also update the time provider above it too

**Comment by Aayush-Abhyarthi** [comment]
To enable elser model, "`platinum`" plan will have to be used. I did the changes in the DA to use the "`platinum`" plan. But the test is failing with the error - "`elasticsearchpl is not available on multitenant`"

**Comment by ocofaigh** [comment]
This work was done in https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/pull/223
closing

**Pull request closed without merging**